### ISSUE TITLE
Implementation goals for Perlcon 2019

### ISSUE THREAD

**Issue by masak** [masak/007#481]
(It's possible some of this issue should be organized as a project, but I haven't learned enough about GitHub projects yet.)

## tl;dr

I plan to go to [YAPC::Europe^WPerlcon](https://perlcon.eu/) in Riga, and I plan to speak there. I would personally be very surprised if it wasn't 007 and macros I ended up talking about.

So, working backwards from that talk, there are some goals I would like for 007 to have met at that point.

(I'm not so good with deadlines, or rather, they have proven to be mostly wishful thinking. See #335. No plan of attack survives its first meeting with the enemy &mdash; and by "enemy", I guess I mean "my general lack of momentum" &mdash; but even so, it's better to _have_ a plan and to make goals than not to.)

### [Hit-the-piñata](https://en.wikipedia.org/wiki/Pi%C3%B1ata) 🎉 goals

* [Separate the logic](https://github.com/masak/007/issues/377#issuecomment-468164322) from the Qtypes (#492)
* Refurbished object system
    * Classes (#32)
    * Enums (#265)
* [New runtime](https://github.com/masak/007/issues/377) (replacing the old one)
    * Compilation visualizer (really good during a talk &mdash; in fact, optimized for the talk)
    * Various optimizations
* [Examples cheating](https://github.com/masak/007/issues/324) (which would also allow us to close #194)
* Rudimentary regexes (#174 and #348)

### [Shoot-for-the-stars](https://en.wikipedia.org/wiki/Per_aspera_ad_astra) :star: goals

* [Hacker News completeness](https://github.com/masak/007/issues/335)
* [Massive cheating](https://github.com/masak/007/issues/401#issuecomment-465883705)
* [Contextual macros](https://github.com/masak/007/issues/349)
* Modules
    * Module installer
    * All of it by not cheating
* Complete regex system
    * Grammars
    * [A new parser](https://github.com/masak/007/issues/293)
        * [Better error messages](https://github.com/masak/007/issues/10)
    * [`is parsed`](https://github.com/masak/007/issues/177)
    * Undoing all the cheating (examples + massive) and replacing it with the real deal
* SEND + MORE = MONEY example implemented as separate languages with transformations between them
* No loose ends/"this is unimplemented" left in the documentation
* [More ambitious phrasing](https://github.com/masak/007/issues/420#issuecomment-460524669) in README.
* v1.0
* [Language renamed Alma](https://github.com/masak/007/issues/277)

## More details, please

XXX I expect to come back here and write things out more concretely, including how things depend on each other, and what we expect to be challenging. But the above is a decent start.

**Comment by masak** [comment]
I don't know exactly why, but somehow the metaphor of striking a piñata while blindfolded feels like a very apt one for the 007 project.

**Comment by masak** [comment]
Talk accepted! Yay!

**Comment by vendethiel** [comment]
I should definitely consider going then :)

**Comment by masak** [comment]
@vendethiel I will be sorely disappointed unless you at least consider it 😉

**Comment by masak** [comment]
Closing this one, as PerlCon has now come and gone.

(I still hope to find time to expand on the roadmap for those goals, because it's still a good set of goals even after the conf.)

**Comment by masak** [comment]
## A more detailed roadmap

My current focus on [Bel](https://github.com/masak/bel) leaves me in practice with zero tuits to advance Alma. But I don't consider the Alma project dead, or finished, or stable; it's more like it reached quite far in its Phase I, where Alma was an experimental interpreter. With the improved perspective I now have from taking a few steps away from the project, I see quite clearly that it's necessary for Alma to enter a Phase II, with a couple of big milestones to drive it:

* **New runtime/Compile to bytecode**. I'm thinking a small, well-defined subset of Alma can be implemented in LLVM. As a first step, maybe not even the Qtypes need to be exposed in userspace.
* **Implement the ambitious parser**. As pointed out in #293 and #485, the really interesting uses of macros depend on parser technology which, to be blunt, doesn't seem to exist yet. (I have some personal notes where it simply says "parser handover; parser relinquishing". This is crucial both for DSLs and for syntax extension, and it will probably figure heavily in the unquote story as well.) At this point I don't care if the parser is bootstrapped in Alma itself; that can come later. The important thing is a clear proof-of-concept that this kind of parser works.
* **Object system/MOP**. This one isn't and never was a central part of Alma's mission, but it still feels like an important thing to get into place.
* **Reflected Qtypes, quasis, macros**. Should let us close #335, which is mostly blocking on #30, which is completely blocking on the ambitious parser.
* **Userland regex/grammar story**, which of course should mesh well with `is parsed`, and the Alma grammar.
* **A strong slangs/langs story**.
* **Modules**.
* **Control flow primitives/low-level code** à la #145.

Again, the question is where I'll be able to come by one or more clones of myself with the tuits to do the above, but... there it is: an updated roadmap for Alma.
